src/generic/numberparsing.h

```diff
@@ -1,80 +1,1145 @@
+                                 0x47c6b82ef32a2069,
+                                 0x4cdc331d57fa5441,
+                                 0xe0133fe4adf8e952,
+                                 0x58180fddd97723a6,
+                                 0x570f09eaa7ea7648};
+
+// Attempts to compute i * 10^(power) exactly; and if "negative" is
+// true, negate the result.
+// This function will only work in some cases, when it does not work, success is
+// set to false. This should work *most of the time* (like 99% of the time).
+// We assume that power is in the [FASTFLOAT_SMALLEST_POWER,
+// FASTFLOAT_LARGEST_POWER] interval: the caller is responsible for this check.
+// We assume that i is non-zero: the caller is responsible for this check.
+really_inline double compute_float_64(int64_t power, uint64_t i, bool negative,
+                                      bool *success) {
+  // we start with a fast path
+  if (-22 <= power && power <= 22 && i <= 9007199254740991) {
+    // convert the integer into a double. This is lossless since
+    // 0 <= i <= 2^53 - 1.
+    double d = i;
+    //
+    // The general idea is as follows.
+    // If 0 <= s < 2^53 and if 10^0 <= p <= 10^22 then
+    // 1) Both s and p can be represented exactly as 64-bit floating-point
+    // values
+    // (binary64).
+    // 2) Because s and p can be represented exactly as floating-point values,
+    // then s * p
+    // and s / p will produce correctly rounded values.
+    //
+    if (power < 0) {
+      d = d / power_of_ten[-power];
+    } else {
+      d = d * power_of_ten[power];
+    }
+    if (negative) {
+      d = -d;
+    }
+    *success = true;
+    return d;
+  }
+
+  components c =
+      power_of_ten_components[power - FASTFLOAT_SMALLEST_POWER]; // safe because
+                                                                 // power_index
+                                                                 // <= 2*308
+  // we recover the mantissa of the power, it has a leading 1. It is always
+  // rounded down.
+  uint64_t factor_mantissa = c.mantissa;
+
+  // We want the most significant bit of i to be 1. Shift if needed.
+  int lz = leading_zeroes(i);
+  i <<= lz;
+  // We want the most significant 64 bits of the product. We know
+  // this will be non-zero because the most significant bit of i is
+  // 1.
+  value128 product = full_multiplication(i,factor_mantissa);
+  uint64_t lower = product.low;
+  uint64_t upper = product.high;
+
+  // We know that upper has at most one leading zero because
+  // both i and  factor_mantissa have a leading one. This means
+  // that the result is at least as large as ((1<<63)*(1<<63))/(1<<64).
+
+  // As long as the first 9 bits of "upper" are not "1", then we
+  // know that we have an exact computed value for the leading
+  // 55 bits because any imprecision would play out as a +1, in
+  // the worst case.
+  if (unlikely((upper & 0x1FF) == 0x1FF) && (lower + i < lower)) {
```

Comment:
how often is it that `lower + i < lower` when `(upper & 0x1FF) == 0x1FF`?

Reply:
Frequently. It is much higher than 1% but lower than 50%. Maybe 20% of the time for canada.json.